### ISSUE TITLE
cryptoservice: the SecureRandom instance is never re-created/seeded

### ISSUE THREAD

**Issue by ols-emuller** [jpos/jPOS-EE#92]
CryptoService creates a SecureRandom instance to create random IVs, but it's never re-created or re-seeded. Is that intended?

Bear in mind that a JPOS application could be running for months and performing thousands of encryptions, and articles like this one https://www.synopsys.com/blogs/software-security/proper-use-of-javas-securerandom/ suggest that a SecureRandom instance shouldn't be used for a long time with the same seed. Maybe a simple solution like creating a new instance of SecureRandom at the same time a new AES key is created would be an improvement.

Thank you, and please let me know if there is a better place to ask questions like this.

**Comment by jameshilliard** [comment]
> Bear in mind that a JPOS application could be running for months and performing thousands of encryptions, and articles like this one https://www.synopsys.com/blogs/software-security/proper-use-of-javas-securerandom/ suggest that a SecureRandom instance shouldn't be used for a long time with the same seed.

This information is incorrect, a properly designed CSPRNG will never run out of entropy and effectively gets more secure over time as additional entropy is mixed in to the pool, Linux does this automatically for the default `/dev/urandom` `NativePRNG` provider for `SecureRandom()`. Keep in mind 256 bits of true random entropy is all you need to securely generate random numbers indefinitely even if you never add additional entropy to the pool.

**Comment by ols-emuller** [comment]
> > Bear in mind that a JPOS application could be running for months and performing thousands of encryptions, and articles like this one https://www.synopsys.com/blogs/software-security/proper-use-of-javas-securerandom/ suggest that a SecureRandom instance shouldn't be used for a long time with the same seed.
> 
> This information is incorrect, a properly designed CSPRNG will never run out of entropy and effectively gets more secure over time as additional entropy is mixed in to the pool, Linux does this automatically for the default `/dev/urandom` `NativePRNG` provider for `SecureRandom()`. Keep in mind 256 bits of true random entropy is all you need to securely generate random numbers indefinitely even if you never add additional entropy to the pool.

Hi! My only question about that is: Is that true for MS Windows too? (*). Apart from that, I only had that article as hint, and the openjdk 8 source code that I tried to follow but gave up at some point, so I can't really argue .

(*)  And now I'm not really sure what secure random generator is being selected on Windows by default when doing `new SecureRandom()`.

**Comment by jameshilliard** [comment]
I think by default it uses the `DRBG` RNG implementation for windows.

**Comment by ols-emuller** [comment]
Just FYI, I just found this article https://www.veracode.com/blog/research/cryptographically-secure-pseudo-random-number-generator-csprng

**Comment by ols-emuller** [comment]
> I think by default it uses the `DRBG` RNG implementation for windows.

.. and according to that veracode article: _On Windows, the default implementation will return the SHA1PRNG algorithm (assuming default configuration of java.security)_

**Comment by jameshilliard** [comment]
Yeah, information seems to vary a bit but luckily either should be good enough for us. The known vulnerabilities affecting SHA1 are collision attacks which don't really affect PRNG use cases from my understanding so we should be fine.

**Comment by ols-emuller** [comment]
ok, thanks!
